File: advent/2021/23.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"os"
	"time"
)

// There are 23 valid cells, so let's get all of them indexed
var cellsRow = [23]int{0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4}
var cellsCol = [23]int{1, 2, 4, 6, 8, 10, 11, 3, 3, 3, 3, 5, 5, 5, 5, 7, 7, 7, 7, 9, 9, 9, 9}

type State struct {
	// Each element represents the state of a cell.
	// A, B, C, D indicate the amphipods while the period mark indicates empty cell
	cells []byte
	// Total cost to reach this state from initial state
	cost int
}

type Move struct {
	// Indices of cells where amphipod stays before and after moving
	orig, dest int
	// Cost of a single move
	cost int
}

func newState(grid [][]byte) State {
	maxRow := len(grid) - 2
	cells := make([]byte, 23)
	for i := range cells {
		x, y := cellsRow[i], cellsCol[i]
		if x > 0 && x <= maxRow {
			cells[i] = grid[x][y]
		} else {
			cells[i] = '.'
		}
	}
	return State{cells, 0}
}

// Weighted costs by amphipod type
var costs = map[byte]int{
	'A': 1,
	'B': 10,
	'C': 100,
	'D': 1000,
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

// Returns information about one amphipod's move
func (s *State) getMove(index, destRow, destCol int) Move {
	cell := s.cells[index]
	x, y := cellsRow[index], cellsCol[index]
	dist := abs(x-destRow) + abs(y-destCol)
	cost := dist * costs[cell]
	destIndex := getCellIndex(destRow, destCol)
	return Move{index, destIndex, cost}
}

// Creates the resulting state after applying one move on original state and returns it
func (s *State) applyMove(m Move) State {
	cellsTemp := make([]byte, len(s.cells))
	copy(cellsTemp, s.cells)
	cellsTemp[m.dest] = cellsTemp[m.orig]
	cellsTemp[m.orig] = '.'
	return State{cellsTemp, s.cost + m.cost}
}

// Checks if amphipod can move along the hallway starting from column x to destX
func (s *State) isHallwayOpen(col, destCol, index int) bool {
	lCol, rCol := col, destCol
	if lCol > rCol {
		lCol, rCol = rCol, lCol
	}
	for i, cell := range s.cells {
		x, y := cellsRow[i], cellsCol[i]
		if i != index && x == 0 && y >= lCol && y <= rCol && cell != '.' {
			return false
		}
	}
	return true
}

var digits = map[byte]int64{
	'.': 0,
	'A': 1,
	'B': 2,
	'C': 3,
	'D': 4,
}

// Since the cell state can only be one of 5 values (A, B, C, D, empty), the cells
// attribute can be represented in a base 5 system, of which the largest value
// (5 pow 23 - 1 ~ 1.192e16) is within int64 range
func (s *State) cellsHash() int64 {
	var key int64
	for _, cell := range s.cells {
		key = 5*key + digits[cell]
	}
	return key
}

func getCellIndex(x, y int) int {
	for i := 0; i < 23; i++ {
		if cellsRow[i] == x && cellsCol[i] == y {
			return i
		}
	}
	return -1
}

// Returns a cell at given coordinate, if a cell is out of range, `\0` is returned
func (s *State) getCell(x, y int) byte {
	index := getCellIndex(x, y)
	if index == -1 {
		return 0
	}
	return s.cells[index]
}

func (s *State) amphipodCount() int {
	count := 0
	for _, cell := range s.cells {
		if cell != '.' {
			count++
		}
	}
	return count
}

var destRoomCol = map[byte]int{
	'A': 3,
	'B': 5,
	'C': 7,
	'D': 9,
}

func (s *State) availableMoves() []Move {
	moves := []Move{}
	maxRow := s.amphipodCount() / 4

nextCell:
	for i, cell := range s.cells {
		// ignore empty cells
		if cell == '.' {
			continue
		}
		col, row := cellsCol[i], cellsRow[i]
		destCol := destRoomCol[cell]

		if row == 0 {
			// Hallway to target room transition:
			// Iterate over the target room from the bottom up to find the target cell.
			// Check if there are amphipods of other type inside the room
			// so they won't get stuck when the amphipod moves in target room
			destRow := 0
			for x := maxRow; x >= 1; x-- {
				c := s.getCell(x, destCol)
				if c == '.' {
					destRow = x
					break
				} else if c != cell {
					continue nextCell
				}
			}
			// Also, make sure the hallway segment between the amphipod and the
			// target room is empty before moving
			if s.isHallwayOpen(col, destCol, i) {
				moves = append(moves, s.getMove(i, destRow, destCol))
			}
		} else {
			// Room to hallway transition:
			// Amphipod can move to the hallway when there are no amphipods overhead
			for x := 1; x < row; x++ {
				if s.getCell(x, col) != '.' {
					continue nextCell
				}
			}
			// If it's already in the target room, it only moves out when it has
			// to make way for amphipods underneath that are in the wrong room to
			// move to their target rooms
			if col == destCol {
				isSet := true
				for x := row + 1; x <= maxRow; x++ {
					if s.getCell(x, col) != cell {
						isSet = false
						break
					}
				}
				if isSet {
					continue
				}
			}
			// Move to all reachable cells in the hallway
			for _, hallCol := range []int{1, 2, 4, 6, 8, 10, 11} {
				if s.isHallwayOpen(col, hallCol, i) {
					moves = append(moves, s.getMove(i, 0, hallCol))
				}
			}
		}
	}
	return moves
}

// Two states are considered equal when the cell states of both are the same
func (s *State) isEqualTo(other State) bool {
	if len(s.cells) != len(other.cells) {
		return false
	}
	for i := 0; i < len(s.cells); i++ {
		if s.cells[i] != other.cells[i] {
			return false
		}
	}
	return true
}

// Priority queue implementation
type PQueue []*State

func (h PQueue) Len() int           { return len(h) }
func (h PQueue) Less(i, j int) bool { return h[i].cost < h[j].cost }
func (h PQueue) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *PQueue) Push(x any)        { *h = append(*h, x.(*State)) }

func (h *PQueue) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func dijkstra(init, end State) (int, []*State) {
	costs := make(map[int64]int)
	prevs := make(map[int64]*State)
	pq := PQueue{&init}
	heap.Init(&pq)

	for len(pq) > 0 {
		state := heap.Pop(&pq).(*State)
		// If end state is reached, trace state path and return total cost
		if state.isEqualTo(end) {
			path := []*State{}
			cur := state
			for cur.isEqualTo(init) {
				path = append(path, cur)
				cur = prevs[cur.cellsHash()]
			}
			path = append(path, &init)
			for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
				path[i], path[j] = path[j], path[i]
			}
			return state.cost, path
		}

		for _, move := range state.availableMoves() {
			next := state.applyMove(move)
			nextKey := next.cellsHash()
			if c, ok := costs[nextKey]; !ok || next.cost < c {
				prevs[nextKey] = state
				costs[nextKey] = next.cost
				heap.Push(&pq, &next)
			}
		}
	}
	return -1, nil
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()

	grid := [][]byte{}
	for scanner.Scan() {
		row := []byte(scanner.Text())
		grid = append(grid, row)
	}
	init1 := newState(grid)
	end1 := State{
		cells: []byte(".......AA..BB..CC..DD.."),
		cost:  0,
	}

	// Add 2 more lines in the input for part 2
	grid = append(grid[:4], grid[2:]...)
	grid[2] = []byte("  #D#C#B#A#")
	grid[3] = []byte("  #D#B#A#C#")
	init2 := newState(grid)
	end2 := State{
		cells: []byte(".......AAAABBBBCCCCDDDD"),
		cost:  0,
	}

	startTime := time.Now()
	p1Cost, _ := dijkstra(init1, end1)
	p1Time := fmt.Sprint(time.Since(startTime))

	startTime = time.Now()
	p2Cost, _ := dijkstra(init2, end2)
	p2Time := fmt.Sprint(time.Since(startTime))

	fmt.Printf("Part 1: %d (%s elapsed)\n", p1Cost, p1Time)
	fmt.Printf("Part 2: %d (%s elapsed)\n", p2Cost, p2Time)
}

File: advent/2021/23.rs
use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap};
use std::io::{self, BufRead};
use std::time::Instant;

const CELLS_ROW: [i8; 23] = [
    0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4,
];
const CELLS_COL: [i8; 23] = [
    1, 2, 4, 6, 8, 10, 11, 3, 3, 3, 3, 5, 5, 5, 5, 7, 7, 7, 7, 9, 9, 9, 9,
];

fn cell_coord(idx: usize) -> (i8, i8) {
    (CELLS_ROW[idx], CELLS_COL[idx])
}

fn cell_index_at_coord(coord: (i8, i8)) -> Option<usize> {
    (0..23).find(|&i| CELLS_ROW[i] == coord.0 && CELLS_COL[i] == coord.1)
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Cell {
    E = 0,
    A = 1,
    B = 2,
    C = 3,
    D = 4,
}

impl From<char> for Cell {
    fn from(c: char) -> Self {
        match c {
            'A' => Cell::A,
            'B' => Cell::B,
            'C' => Cell::C,
            'D' => Cell::D,
            '.' => Cell::E,
            c => panic!("invalid char {:?}", c),
        }
    }
}

impl Cell {
    fn target_column(&self) -> i8 {
        const TARGET_COLS: [i8; 5] = [0, 3, 5, 7, 9];
        match self {
            Cell::E => panic!("guard this case yourself"),
            c => TARGET_COLS[*c as usize],
        }
    }

    fn weight(&self) -> u32 {
        const WEIGHTS: [u32; 5] = [0, 1, 10, 100, 1000];
        WEIGHTS[*self as usize]
    }
}

#[derive(Clone, Debug)]
struct State {
    cells: Vec<Cell>,
    cost: u32,
}

#[derive(Clone, Debug)]
struct Move {
    orig_idx: usize,
    dest_idx: usize,
    cost: u32,
}

impl State {
    fn parse_grid(grid: Vec<Vec<char>>) -> Self {
        let room_depth = grid.len() as i8 - 2;
        let mut cells = vec![Cell::E; 23];
        for (i, c) in cells.iter_mut().enumerate() {
            let (x, y) = cell_coord(i);
            if x > 0 && x <= room_depth {
                *c = Cell::from(grid[x as usize][y as usize]);
            }
        }
        State { cells, cost: 0 }
    }

    // Since a cell can be one of 5 values (A, B, C, D, empty), the cells attribute
    // can be hashed in a base 5 system number
    fn cells_hash(&self) -> u64 {
        self.cells.iter().fold(0, |hash, &c| hash * 5 + c as u64)
    }

    fn amphipod_count(&self) -> usize {
        self.cells.iter().filter(|&&x| x != Cell::E).count()
    }

    fn get_move(&self, u_idx: usize, v: (i8, i8)) -> Option<Move> {
        assert!(u_idx < 23);
        let u = cell_coord(u_idx);
        let dist = (v.0 - u.0).abs() + (v.1 - u.1).abs();
        cell_index_at_coord(v).map(|dest_idx| Move {
            orig_idx: u_idx,
            dest_idx,
            cost: dist as u32 * self.cells[u_idx].weight(),
        })
    }

    fn apply_move(&self, m: &Move) -> State {
        let mut cells = self.cells.clone();
        cells[m.dest_idx] = cells[m.orig_idx];
        cells[m.orig_idx] = Cell::E;
        State {
            cells,
            cost: self.cost + m.cost,
        }
    }

    fn is_hallway_segment_empty(&self, idx: usize, col_a: i8, col_b: i8) -> bool {
        let col_range = col_a.min(col_b)..=col_a.max(col_b);
        (0..self.cells.len())
            .filter(|&i| {
                let (r, c) = cell_coord(i);
                r == 0 && col_range.contains(&c) && i != idx
            })
            .all(|i| self.cells[i] == Cell::E)
    }

    fn cell_at_coord(&self, coord: (i8, i8)) -> Option<Cell> {
        cell_index_at_coord(coord).map(|idx| self.cells[idx])
    }

    fn available_moves(&self) -> Vec<Move> {
        let mut moves = vec![];
        let row_max = self.amphipod_count() as i8 / 4;

        'next_cell: for (i, &cell) in self.cells.iter().enumerate() {
            if cell == Cell::E {
                continue;
            }
            let (row, col) = cell_coord(i);

            if row == 0 {
                // Hallway to target room transition
                let dest_col = cell.target_column();
                let mut dest_row = 0;

                // Iterate over the target room from the bottom up to find the target cell.
                // Check if there are amphipods of other type inside the room
                // so they won't get stuck when the amphipod moves in target room
                for x in (1..=row_max).rev() {
                    match self.cell_at_coord((x, dest_col)) {
                        Some(Cell::E) => {
                            dest_row = x;
                            break;
                        }
                        Some(c) if c != cell => continue 'next_cell,
                        Some(_) => (),
                        None => panic!("how is {:?} not a cell", (x, dest_col)),
                    };
                }

                // Also, make sure the hallway segment between the amphipod and the
                // target room is empty before moving
                assert_ne!(dest_row, 0);
                if self.is_hallway_segment_empty(i, col, dest_col) {
                    moves.push(self.get_move(i, (dest_row, dest_col)).unwrap());
                }
            } else {
                // Room to hallway transition
                // Amphipod can move to the hallway when there are no amphipods overhead
                for x in 1..row {
                    match self.cell_at_coord((x, col)) {
                        Some(Cell::E) => (),
                        Some(_) => continue 'next_cell,
                        None => panic!("how is {:?} not a cell", (x, col)),
                    };
                }

                // If it's already in the target room, it only moves out when it has
                // to make way for amphipods underneath that are in the wrong room to
                // move to their target rooms
                if col == cell.target_column() {
                    let mut all_comrades = true;
                    for x in row + 1..=row_max {
                        match self.cell_at_coord((x, col)) {
                            Some(Cell::E) => panic!("they're not gonna bite you bro"),
                            Some(c) if c != cell => {
                                all_comrades = false;
                                break;
                            }
                            Some(_) => (),
                            None => panic!("how is {:?} not a cell", (x, col)),
                        }
                    }
                    if all_comrades {
                        continue;
                    }
                }

                const HALLWAY_COLS: [i8; 7] = [1, 2, 4, 6, 8, 10, 11];
                for &y in HALLWAY_COLS.iter() {
                    if self.is_hallway_segment_empty(i, col, y) {
                        moves.push(self.get_move(i, (0, y)).unwrap());
                    }
                }
            }
        }
        moves
    }

    fn equal_cells_with(&self, other: &Self) -> bool {
        self.cells
            .iter()
            .zip(other.cells.iter())
            .all(|(a, b)| a == b)
    }
}

impl PartialEq for State {
    fn eq(&self, other: &Self) -> bool {
        self.cost == other.cost
    }
}

impl Eq for State {}

impl Ord for State {
    fn cmp(&self, other: &Self) -> Ordering {
        other.cost.cmp(&self.cost)
    }
}

impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn dijkstra(init: &State, end: &State) -> Option<(Vec<Move>, u32)> {
    let mut costs: HashMap<u64, u32> = HashMap::new();
    let mut prevs: HashMap<u64, (u64, Move)> = HashMap::new();

    let mut queue = BinaryHeap::new();
    queue.push(init.clone());

    while let Some(state) = queue.pop() {
        // if end state is reached, trace move path and return total cost
        if state.equal_cells_with(end) {
            let mut path = vec![];
            let init_hash = init.cells_hash();
            let mut hash = state.cells_hash();
            while hash != init_hash {
                let (h, m) = prevs.get(&hash).unwrap();
                path.push(m.clone());
                hash = *h;
            }
            path.reverse();
            return Some((path, state.cost));
        }

        let moves = state.available_moves();
        for m in moves.iter() {
            let next = state.apply_move(m);
            let next_key = next.cells_hash();
            match costs.get(&next_key) {
                Some(&c) if next.cost >= c => continue,
                _ => {
                    costs.insert(next_key, next.cost);
                    prevs.insert(next_key, (state.cells_hash(), m.clone()));
                }
            };
            queue.push(next);
        }
    }
    None
}

fn main() {
    use Cell::*;

    let mut lines = io::stdin().lock().lines().map(Result::unwrap);
    lines.next();
    let grid1: Vec<Vec<_>> = lines.map(|l| l.chars().collect()).collect();
    let mut grid2 = grid1.clone();

    let init1 = State::parse_grid(grid1);
    let end1 = State {
        cells: vec![
            E, E, E, E, E, E, E, A, A, E, E, B, B, E, E, C, C, E, E, D, D, E, E,
        ],
        cost: 0,
    };

    // add 2 more lines for part 2
    grid2.insert(2, "  #D#C#B#A#".chars().collect());
    grid2.insert(3, "  #D#B#A#C#".chars().collect());
    let init2 = State::parse_grid(grid2);
    let end2 = State {
        cells: vec![
            E, E, E, E, E, E, E, A, A, A, A, B, B, B, B, C, C, C, C, D, D, D, D,
        ],
        cost: 0,
    };

    let start = Instant::now();
    let (_p, c) = dijkstra(&init1, &end1).unwrap();
    let p1_time = start.elapsed();
    // for m in p.iter() {
    //     println!("{:?} -> {:?} ", cell_coord(m.orig_idx), cell_coord(m.dest_idx));
    // }
    println!("Part 1: {} ({}ms elapsed)", c, p1_time.as_millis());
    let start = Instant::now();
    let (_p, c) = dijkstra(&init2, &end2).unwrap();
    let p2_time = start.elapsed();
    // for m in p.iter() {
    //     println!("{:?} -> {:?} ", cell_coord(m.orig_idx), cell_coord(m.dest_idx));
    // }
    println!("Part 2: {} ({}ms elapsed)", c, p2_time.as_millis());
}
